src/cpp/fastdds/domain/DomainParticipantImpl.cpp

```diff
@@ -1532,10 +1531,17 @@ void DomainParticipantImpl::MyRTPSParticipantListener::onParticipantDiscovery(
         RTPSParticipant*,
         ParticipantDiscoveryInfo&& info)
 {
-    DomainParticipantListener* listener = nullptr;
-    if (participant_ != nullptr && (listener = participant_->get_listener()) != nullptr)
+    Sentry sentinel(this);
+    if (sentinel)
     {
-        listener->on_participant_discovery(participant_->participant_, std::move(info));
+        DomainParticipantListener* listener = nullptr;
```

Comment:
We no longer need local variables. If we have races we want TSAN to report them. We can use `participant_->get_listener()` confidently because the `sentry` is on guard.

Reply:
I simply say that doing:
```
    Sentry sentinel(this);
    if (sentinel)
    {
        participant_->get_participant()->get_listener()->on_participant_discovery(participant, std::move(info));
    }
```
is perfectly safe because of `Sentry` magic.

Reply:
I removed the local variables and added two new conditions to `SEntry` constructor.